tempesta_fw/http_msg.c

```diff
@@ -1462,17 +1434,30 @@ tfw_h2_msg_rewrite_data(TfwHttpTransIter *mit, const TfwStr *str,
 /**
  * Insert data from string @data to message at offset defined by message
  * iterator @it and @off. This function doesn't maintain message structure.
- * After insertion message iterator will point at the end of fragment right
- * before the inserted data.
+ * After insertion message iterator and @data will point at the start of
+ * inserted data fragment.
  */
 int
-tfw_http_msg_insert(TfwMsgIter *it, char *off, const TfwStr *data)
+tfw_http_msg_insert(TfwMsgIter *it, char **off, const TfwStr *data)
 {
 	int r;
 	TfwStr dst = {};
 
-	if ((r = ss_skb_get_room(it->skb_head, it->skb, off, data->len, &dst)))
+	if ((r = ss_skb_get_room(it->skb_head, it->skb, *off, data->len, &dst)))
 		return r;
 
+	/*
+	 * @dst is always a plain string. If the @data wasn't inserted into
+	 * current skb fragment, then it was appended to a previous one.
+	 * Insertions into linear part doesn't require iterator updates.
+	 */
+	*off = dst.data;
+	if (it->frag >= 0) {
+		skb_frag_t *f = &skb_shinfo(it->skb)->frags[it->frag];
+		long offset = dst.data - (char *)skb_frag_address(f);
+		if (offset < 0 || offset > skb_frag_size(f))
+			it->frag--;
+	}
```

Comment:
Didn't think about it as move back. Just thought about is as the laziest and assumption-free way to obtain valid `(it, off)` pair:  `ss_skb_get_room()` can split skbs, if there are too many fragments, and the original iterator fragment may be moved to a next skb. Happens quite frequently on large bodies proxying. 

The `ss_skb_get_room()` always return plain string with exactly pointed beginning, only fragment number is unclear. It's easy to get a pointer to the first or the last byte of the string. But pointer to the byte just behind the string is bit more complicated: `it->skb`, `it->frag`, `off` - they're all will be updated, and we also need a check, that `it->skb` doesn't point to `it->skb_head`. This semantically equals to a single `iter_move()` call. With this in mind `iter_move()` will be called twice: first to step behind the inserted string, and second  - to skip next `max_sz` bytes.

Since the data is always inserted as a plain string, `iter_move()` will jump over the whole string by one iteration. That's why I see no practical reason to move the iterator in `tfw_http_msg_insert()` behind the inserted string.

Reply:
My point isn't about DRY principle and basically I'm OK with you explanation. My concern is that the current h2 code is overcomplicated with the fragments and I'm afraid that we won't see good enough performance. We should have been benefited from the zero copy approach and avoiding memory allocations, but in fact we do a lot of _senseless_ logic (in general meaning, not the current patch). For example, `tfw_msg_iter_move()` for DATA frame is called after `tfw_http_msg_insert()` for HEADERS and both the functions check `it->frag` for zero. The worst thing about both the functions is that both of them were written to be called from `__tfw_h2_make_frames()` only, which is called from only 2 pieces of the same function `tfw_h2_make_frames()`. So there is generic code with extra conditions (and also complexity!), but I don't see enough motivation for the generic functions.

I ask you to review this place carefully and rework in in simpler and more efficient way. Otherwise we might find ourselves as Nginx: there is a long tail in perf and the whole code must be rewritten to make it more or less faster. We defenitelly won't see these two function in perf top, but we do need to remove such ineficiencies as we observe them.